core/src/main/java/org/infinispan/interceptors/distribution/ScatteredDistributionInterceptor.java

```diff
@@ -272,13 +289,14 @@ private Object handleWriteOnOriginPrimary(InvocationContext ctx, DataWriteComman
       Address backup = getNextMember(cacheTopology);
       if (backup != null) {
          // error responses throw exceptions from JGroupsTransport
-         CompletableFuture<Map<Address, Response>> rpcFuture =
-            rpcManager.invokeRemotelyAsync(Collections.singletonList(backup), backupCommand, defaultSyncOptions);
+         List<Address> recipients = Collections.singletonList(backup);
+         CompletableFuture<?> rpcFuture = rpcManager.invokeRemotelyAsync(recipients, backupCommand, defaultSyncOptions);
          rpcFuture.thenRun(() -> {
             if (cacheEntry.isCommitted() && !command.hasAnyFlag(FlagBitSets.PUT_FOR_STATE_TRANSFER)) {
-               svm.scheduleKeyInvalidation(command.getKey(), cacheEntry.getMetadata().version(), cacheEntry.isRemoved());
+               scheduleKeyInvalidation(command.getKey(), cacheEntry.getMetadata().version(), cacheEntry.isRemoved());
```

Comment:
Accessing the command and cache entry here seems like a really bad idea, because the lambda could run while another interceptor is also modifying the context. IMO CompletableFuture/CompletionStage composition results should only be discarded when there is a very good (and documented) reason to do so.

Reply:
True, it would be better to get the key & version & remove and pass them as the part of the closure.